snapshot? tests are faillings

Hi, nixpkgs (nixos package repository) contributer here. I am currently working on updating the memray package, as requested by on of our user:
- https://github.com/NixOS/nixpkgs/issues/330398

However, in this process, I've encountered some failling tests:
```
===== 13 failed, 508 passed, 8 skipped, 4 deselected, 4 warnings in 54.41s =====
```
```
Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
rootdir: /build/source
configfile: pyproject.toml
plugins: textual-snapshot-0.4.0, syrupy-4.6.1
collected 533 items / 4 deselected / 529 selected                              

tests/integration/test_api.py .......                                    [  1%]
tests/integration/test_attach.py ssssssss                                [  2%]
tests/integration/test_extensions.py .........                           [  4%]
tests/integration/test_greenlet.py ..                                    [  4%]
tests/integration/test_ipython.py ..........                             [  6%]
tests/integration/test_native_tracking.py ...........                    [  8%]
tests/integration/test_processes.py ....                                 [  9%]
tests/integration/test_socket.py ...........                             [ 11%]
tests/integration/test_threads.py ...                                    [ 12%]
tests/integration/test_tracing.py ...........................            [ 17%]
tests/integration/test_tracking.py ..................................... [ 24%]
........................................................................ [ 37%]
........                                                                 [ 39%]
tests/test_utils.py .......                                              [ 40%]
tests/unit/test_allocation_lifetime_aggregator.py ...................    [ 44%]
tests/unit/test_attach.py .                                              [ 44%]
tests/unit/test_cli.py ................................................. [ 53%]
..........................                                               [ 58%]
tests/unit/test_flamegraph_reporter.py ................................. [ 65%]
...                                                                      [ 65%]
tests/unit/test_frame_tools.py .........................                 [ 70%]
tests/unit/test_high_water_mark_aggregator.py .......................... [ 75%]
......                                                                   [ 76%]
tests/unit/test_highwatermark_command.py ..............                  [ 79%]
tests/unit/test_reader.py .....                                          [ 79%]
tests/unit/test_stats_reporter.py ..........                             [ 81%]
tests/unit/test_summary_reporter.py .....                                [ 82%]
tests/unit/test_table_reporter.py .....                                  [ 83%]
tests/unit/test_templates.py ......                                      [ 84%]
tests/unit/test_tracker.py ..                                            [ 85%]
tests/unit/test_transform_reporter.py ...........                        [ 87%]
tests/unit/test_tree_reporter.py ........................FFFFF.F.        [ 93%]
tests/unit/test_tui_reporter.py ......................FFFF...F...FF      [100%]

=================================== FAILURES ===================================
```

I noticed the `Update textual-snapshot files to new format` commit, applied it on top of the `1.13.4` release source, but it also fails:

```
=========== 21 failed, 500 passed, 8 skipped, 4 deselected in 57.72s ===========
```
```
Executing pytestCheckPhase
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
rootdir: /build/source
configfile: pyproject.toml
plugins: textual-snapshot-0.4.0, syrupy-4.6.1, cov-5.0.0
collected 533 items / 4 deselected / 529 selected                              

tests/integration/test_api.py .......                                    [  1%]
tests/integration/test_attach.py ssssssss                                [  2%]
tests/integration/test_extensions.py .........                           [  4%]
tests/integration/test_greenlet.py ..                                    [  4%]
tests/integration/test_ipython.py ..........                             [  6%]
tests/integration/test_native_tracking.py ...........                    [  8%]
tests/integration/test_processes.py ....                                 [  9%]
tests/integration/test_socket.py ...........                             [ 11%]
tests/integration/test_threads.py ...                                    [ 12%]
tests/integration/test_tracing.py ...........................            [ 17%]
tests/integration/test_tracking.py ..................................... [ 24%]
........................................................................ [ 37%]
........                                                                 [ 39%]
tests/test_utils.py .......                                              [ 40%]
tests/unit/test_allocation_lifetime_aggregator.py ...................    [ 44%]
tests/unit/test_attach.py .                                              [ 44%]
tests/unit/test_cli.py ................................................. [ 53%]
..........................                                               [ 58%]
tests/unit/test_flamegraph_reporter.py ................................. [ 65%]
...                                                                      [ 65%]
tests/unit/test_frame_tools.py .........................                 [ 70%]
tests/unit/test_high_water_mark_aggregator.py .......................... [ 75%]
......                                                                   [ 76%]
tests/unit/test_highwatermark_command.py ..............                  [ 79%]
tests/unit/test_reader.py .....                                          [ 79%]
tests/unit/test_stats_reporter.py ..........                             [ 81%]
tests/unit/test_summary_reporter.py .....                                [ 82%]
tests/unit/test_table_reporter.py .....                                  [ 83%]
tests/unit/test_templates.py ......                                      [ 84%]
tests/unit/test_tracker.py ..                                            [ 85%]
tests/unit/test_transform_reporter.py ...........                        [ 87%]
tests/unit/test_tree_reporter.py .....................FFFFFFFFFFF        [ 93%]
tests/unit/test_tui_reporter.py ......................FFFFFFFF...FF      [100%]

=================================== FAILURES ===================================
```

I am not sure what this means for memray usability (are those ttest critical?), so I am opening this issue to get more insight

--- a/requirements-test.txt
+++ b/requirements-test.txt
@@ -8,3 +8,4 @@ setuptools; python_version >= '3.12'
 pkgconfig
 pytest-textual-snapshot
 textual >= 0.43, != 0.65.2, != 0.66
+packaging

--- a/setup.py
+++ b/setup.py
@@ -120,6 +120,7 @@ test_requires = [
     "setuptools; python_version >= '3.12'",
     "pytest-textual-snapshot",
     "textual >= 0.43, != 0.65.2, != 0.66",
+    "packaging",
 ]
 
 benchmark_requires = [

--- a/tests/conftest.py
+++ b/tests/conftest.py
@@ -2,6 +2,12 @@ import socket
 import sys
 
 import pytest
+from packaging import version
+
+SNAPSHOT_MINIMUM_VERSIONS = {
+    "textual": "0.73",
+    "pytest-textual-snapshot": "1.0",
+}
 
 
 @pytest.fixture
@@ -13,7 +19,48 @@ def free_port():
     return port_number
 
 
-if sys.version_info < (3, 8):
-    # Ignore unused Textual snapshots on Python 3.7
-    def pytest_configure(config):
+def _snapshot_skip_reason():
+    if sys.version_info < (3, 8):
+        # Every version available for 3.7 is too old
+        return f"snapshot tests require textual>={SNAPSHOT_MINIMUM_VERSIONS['textual']}"
+
+    from importlib import metadata  # Added in 3.8
+
+    for lib, min_ver in SNAPSHOT_MINIMUM_VERSIONS.items():
+        try:
+            ver = version.parse(metadata.version(lib))
+        except ImportError:
+            return f"snapshot tests require {lib} but it is not installed"
+
+        if ver < version.parse(min_ver):
+            return f"snapshot tests require {lib}>={min_ver} but {ver} is installed"
+
+    return None
+
+
+def pytest_configure(config):
+    if config.option.update_snapshots:
+        from importlib import metadata  # Added in 3.8
+
+        for lib, min_ver in SNAPSHOT_MINIMUM_VERSIONS.items():
+            ver = version.parse(metadata.version(lib))
+            if ver != version.parse(min_ver):
+                pytest.exit(
+                    f"snapshots must be generated with {lib}=={min_ver}"
+                    f" or SNAPSHOT_MINIMUM_VERSIONS must be updated to {ver}"
+                    f" in {__file__}"
+                )
+        return
+
+    reason = _snapshot_skip_reason()
+    if reason:
+        config.issue_config_time_warning(UserWarning(reason), stacklevel=2)
         config.option.warn_unused_snapshots = True
+
+
+def pytest_collection_modifyitems(config, items):
+    reason = _snapshot_skip_reason()
+    if reason:
+        for item in items:
+            if "snap_compare" in item.fixturenames:
+                item.add_marker(pytest.mark.skip(reason=reason))

--- a/tests/unit/test_tree_reporter.py
+++ b/tests/unit/test_tree_reporter.py
@@ -1,4 +1,3 @@
-import sys
 from dataclasses import dataclass
 from textwrap import dedent
 from typing import Any
@@ -1534,15 +1533,6 @@ class TestTreeTui:
 
 @pytest.fixture
 def compare(monkeypatch, tmp_path, snap_compare):
-    # The snapshots we've generated using current versions of Textual aren't
-    # expected to match anymore on Python 3.7, as Textual dropped support for
-    # Python 3.7 in the 0.44 release. However, we'd still like to run our
-    # snapshot tests on Python 3.7, to confirm that no unexpected exceptions
-    # occur and that the app doesn't crash. So, allow `snap_compare()` to drive
-    # the application, but always return `True` on Python 3.7 as long as no
-    # exception was raised.
-    succeed_even_if_mismatched = sys.version_info < (3, 8)
-
     def compare_impl(
         allocations: Iterator[AllocationRecord],
         press: Iterable[str] = (),
@@ -1561,14 +1551,11 @@ def compare(monkeypatch, tmp_path, snap_compare):
         with monkeypatch.context() as app_patch:
             app_patch.setitem(globals(), app_global, app)
             tmp_main.write_text(f"from {__name__} import {app_global} as app")
-            return (
-                snap_compare(
-                    str(tmp_main),
-                    press=press,
-                    terminal_size=terminal_size,
-                    run_before=run_before,
-                )
-                or succeed_even_if_mismatched
+            return snap_compare(
+                str(tmp_main),
+                press=press,
+                terminal_size=terminal_size,
+                run_before=run_before,
             )
 
     yield compare_impl

--- a/tests/unit/test_tui_reporter.py
+++ b/tests/unit/test_tui_reporter.py
@@ -1,6 +1,5 @@
 import asyncio
 import datetime
-import sys
 from io import StringIO
 from typing import Awaitable
 from typing import Callable
@@ -103,15 +102,6 @@ class MockReader:
 def compare(monkeypatch, tmp_path, snap_compare):
     monkeypatch.setattr(memray.reporters.tui, "datetime", FakeDatetime)
 
-    # The snapshots we've generated using current versions of Textual aren't
-    # expected to match anymore on Python 3.7, as Textual dropped support for
-    # Python 3.7 in the 0.44 release. However, we'd still like to run our
-    # snapshot tests on Python 3.7, to confirm that no unexpected exceptions
-    # occur and that the app doesn't crash. So, allow `snap_compare()` to drive
-    # the application, but always return `True` on Python 3.7 as long as no
-    # exception was raised.
-    succeed_even_if_mismatched = sys.version_info < (3, 8)
-
     def compare_impl(
         cmdline_override: Optional[str] = None,
         press: Iterable[str] = (),
@@ -138,14 +128,11 @@ def compare(monkeypatch, tmp_path, snap_compare):
         with monkeypatch.context() as app_patch:
             app_patch.setitem(globals(), app_global, app)
             tmp_main.write_text(f"from {__name__} import {app_global} as app")
-            return (
-                snap_compare(
-                    str(tmp_main),
-                    press=press,
-                    terminal_size=terminal_size,
-                    run_before=run_before_wrapper,
-                )
-                or succeed_even_if_mismatched
+            return snap_compare(
+                str(tmp_main),
+                press=press,
+                terminal_size=terminal_size,
+                run_before=run_before_wrapper,
             )
 
     yield compare_impl

